range-requests/src/lib.rs

```diff
@@ -0,0 +1,441 @@
+    fn eq(&self, other: &Self) -> bool {
+        self.range.start == other.range.start
+            && self.range.length == other.range.length
+            && self.total == other.total
+    }
+}
+
+impl SingleRange {
+    pub fn new(
+        range: http_range::HttpRange,
+        total: u64,
+    ) -> Result<Self, Error> {
+        let http_range::HttpRange { start, mut length } = range;
+
+        const INVALID_RANGE: Error =
+            Error::Parse(http_range::HttpRangeParseError::InvalidRange);
+
+        // Clip the length to avoid going beyond the end of the total range
+        if start.checked_add(length).ok_or(INVALID_RANGE)? >= total {
+            length = total.checked_sub(start).ok_or(INVALID_RANGE)?;
+        }
+        // If the length is zero, we cannot satisfy the range request
+        if length == 0 {
+            return Err(INVALID_RANGE);
+        }
+
+        Ok(Self { range: http_range::HttpRange { start, length }, total })
+    }
+
+    /// Return the first byte in this range for use in inclusive ranges.
+    pub fn start(&self) -> u64 {
+        self.range.start
+    }
+
+    /// Return the last byte in this range for use in inclusive ranges.
+    pub fn end_inclusive(&self) -> u64 {
+        assert!(self.range.length > 0);
+
+        self.range
+            .start
+            .checked_add(self.range.length)
+            .unwrap()
+            .checked_sub(1)
+            .unwrap()
+    }
+
+    /// Generate the Content-Range header for inclusion in a HTTP 206 partial
+    /// content response using this range.
+    pub fn to_content_range(&self) -> String {
+        format!(
+            "bytes {}-{}/{}",
+            self.range.start,
+            self.end_inclusive(),
+            self.total
+        )
+    }
+
+    /// Generate a Range header for inclusion in another HTTP request; e.g.,
+    /// to a backend object store.
+    pub fn to_range(&self) -> String {
+        format!("bytes={}-{}", self.range.start, self.end_inclusive())
+    }
+
+    pub fn content_length(&self) -> u64 {
+        assert!(self.range.length > 0);
+
+        self.range.length
+    }
+}
+
+/// A trait, implemented for [dropshot::RequestContext], to pull a range header
+/// out of the request headers.
+pub trait RequestContextEx {
+    fn range(&self) -> Option<PotentialRange>;
+}
+
+impl<T> RequestContextEx for dropshot::RequestContext<T>
+where
+    T: Send + Sync + 'static,
+{
+    /// If there is a Range header, return it for processing during response
+    /// generation.
+    fn range(&self) -> Option<PotentialRange> {
+        self.request
+            .headers()
+            .get(hyper::header::RANGE)
+            .map(|hv| PotentialRange(hv.as_bytes().to_vec()))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use bytes::Bytes;
+    use futures::stream::once;
+    use http_body_util::BodyExt;
+    use std::convert::Infallible;
+    use tokio_util::io::ReaderStream;
+
+    #[test]
+    fn parse_range_valid() {
```

Comment:
not a blocker, but this feels  like it could potentially be a nice use-case for property tests eventually?

Reply:
Heh I checked to see if omicron pulled in quick check so that I could suggest the same thing and then I realized the crate I actually wanted to suggest was proptest.

Reply:
I went ahead and added a couple proptest tests -- this is my first real usage of it, so I welcome extra scrutiny here.